### DISCUSSION TITLE
عدم تطابق کد رهگیری

### DISCUSSION THREAD

**Discussion by masoud86** [Sina-Soltani/Parbad#211] [Q&A] [answered]
سلام
مشکلی که باهاش مواجه شدم در پرداخت با درگاه بانک پاسارگاد بود که همونطور که توی عکس مشخصه کد رهگیری که بانک توی درگاه نشون میده 987830 هست ولی بعد از دریافت پاسخ بانک پراپرتی
TrackingNumber
توی اینترفیس
IPaymentResult
با مقدار شماره ارجاع داخلی پر میشه
![image](https://user-images.githubusercontent.com/4735034/120062337-f1c7a900-c076-11eb-80c7-dca1620ad44c.png)

خروجی دریافتی از بانک:
![image](https://user-images.githubusercontent.com/4735034/120062545-0c4e5200-c078-11eb-9215-b9f07bc7b4bf.png)

**Comment by Sina-Soltani** [comment]
درود

در تصویری که فرستادید
TrackingNumber
که در خروجی به صورت جیسون گذاشتید برابر هست با
2898516819137325600

طبق گفته شما، این مقدار با شماره ارجاع داخلی پر شده
شماره ارجاع در تصویر اول برابر هست با
۱۴۱۱۱۹۶۳۰۷۲۹

این دو مقدار با هم یکسان نیستن

**Reply by masoud86** [reply]
تصویر اول داخل درگاه پرداخت بانک هست و تصویر دوم مقدار دریافتی در پروژه
تنها مقدار مشترک بین بانک و مقدار دریافتی شماره ارجاع داخلی **(و نه شماره ارجاع)** هست که به عنوان
TransactionCode
مقداردهی میشه
در صورتیکه پراپرتی
TrackingNumber
دریافتی باید با مقدار شماره پیگیری مقداردهی بشه

**Reply by Sina-Soltani** [reply]
با توجه به اینکه (احتمالا) شما اطلاعات پرداخت رو در پایگاه داده خود ذخیره کردید، لطفا چک کنید که آیا
TrackingNumber
در مرحله اول که در پاسخ متد ریکوئست هست، برابر هست با
TrackingNumber
که در مرحله آخر دریافت کردید یا خیر

اگر در پایگاه داده ذخیره نکردید میتونید
Log
کنید تا مطمئن بشید
نتیجه رو لطفا اطلاع بدید تا بررسی کنیم

**Reply by masoud86** [reply]
مقدار ذخیره شده در دیتابیس رو بررسی کردم
مقدار 
TrackingNumber
دریافتی در مرحله 
Request 
نه با مقدار 
TransactionCode 
و نه با مقدار 
TrackingNumber 
در مرحله
Verify
یکسان نبود

**Comment by Sina-Soltani** [comment]
احتمال اینکه 
TrackingNumber
ها برابر نباشند، صفر هست

آیا ذخیره و بازیابی رو خودتون پیاده سازی کردید؟
اگر بله، لطفا چک که مقدار دهی پراپرتی ها رو درست انجام دادید یا خیر

**Reply by masoud86** [reply]
ممنون از پاسختون
چشم حتماً مجدد یکبار دیگه تمام کدها رو چک میکنم اما مشکل اصلی عدم ارسال کد پیگیری 6 رقمی درگاه بانک هست
بعد از بررسی مجدد حتماً نتیجه رو اعلام میکنم

**Reply by masoud86** [reply]
من مجدد کدها رو بررسی کردم
طبق گفته خودتون مقدار 
TrackingCode 
در هر دو مرحله یکسان هست و منظور من هم از متفاوت بودن مقدار کد رهگیری مربوط به مقدار نمایش داده شده در صفحه بانک و مقدار دریافت شده در برنامه بود.
برای روشن شدن مشکل لطفا این تصویر رو مجدد بررسی کنید
![compare](https://user-images.githubusercontent.com/4735034/120071213-7a5c3e80-c0a3-11eb-978d-902126722099.jpg)
تصویر سمت راست مقدار دریافت شده در برنامه و تصویر سمت چپ نتیجه پرداخت در صفحه درگاه (شاپرک) هست
مقداری که به عنوان کد رهگیری به کاربر نمایش داده میشه عدد 987830 هست
اما این مقدار رو توی مدل دریافتی نداریم

**Comment by Sina-Soltani** [comment, marked as answer]
درگاه های بانکی، از نام های متفاوتی در سیستم هاشون استفاده میکنن.
اگر بانکی از نامی استفاده کرد که شاید با نامی در سیستم پَرباد همنام بود، دلیل بر این نیست که اون پراپرتی دقیقا معادل یک پراپرتی در پَرباد هست
خیلی از درگاه ها چندین پراپرتی و شماره های مختلف دیگه ای در سیستم داخلی خودشون دارن که پذیرنده اصولا نیازی به اونها نداره

در واقع اگر بخوام خلاصه تر بگم، در تصویر چپ، اون مقداری که با نام شماره پیگیری نوشته شده، ممکنه هیچ ارتباطی با
TrackingNumber
در سیستم پَرباد نداشته باشه و صرفا چون نام هاشون شبیه به یکدیگر هست دلیل بر برابر بودن اون دو نیست

در سیستم پَرباد
TrackingNumber
معادل شماره پیگیری هست
این مقدار توسط فروشنده تولید میشه و نه درگاه بانکی
آیا عدد ۹۸۷۸۳۰ توسط شما تولید شده و در هنگام ارسال درخواست پرداخت، این عدد رو شما تولید کردید؟

**Reply by masoud86** [reply]
ممنون از پاسخ کاملتون
خیر این عدد رو من تولید نکردم و خود بانک نمایش میده
فقط به این دلیل که داخل درگاه بانک این عدد رو به عنوان کد رهگیری نمایش میده و کد رهگیری ما در وب سایت متفاوت هست برای کاربران سایت ابهام بوجود میاد
اما متوجه صحبت شما هستم و دنبال روشی بودم که شاید بشه همه مواردی که بانک توی درگاهش نشون میده رو در زمان
verify
درخواست داشته باشیم. مثل شماره کارت و یا هر داده دیگه

**Reply by Sina-Soltani** [reply]
TrackingNumber
یک شماره پیگیری هست که توسط فروشنده تولید میشه و باید هنگام درخواست پرداخت، این شماره به درگاه بانک ارسال بشه

تمام درگاه های بانکی این شماره رو باید از فروشنده دریافت کنن. در غیر اینصورت حتی درخواست پرداخت هم نمیتونه انجام بشه چه برسه به اینکه کاربر (خریدار) به درگاه بانکی هدایت بشه و یا پرداختی بخواد صورت بگیره

نمایش دادن شماره پیگیری (که توسط شما تولید شده) به کاربر، یک تصمیم خصوصی برای فروشنده هست. میتونید به خریدار نمایش بدید. میتونید هم نمایش ندید. چون این عدد صرفا جهت عملیات داخلی بانکی استفاده میشه.
شما میتونید حتی به جای نمایش شماره پیگیری (که توسط شما تولید شده)، یک شماره دیگری مانند شماره سفارش به خریدار نمایش بدید
همونطور که گفتم این یک تصمیم داخلی و خصوصی برای فروشنده هست که به چه صورتی تمایل داره بیزینس خود رو اداره کنه

در سیستم بانکی، تنها مقداری که فروشنده باید حتما به خریدار نمایش بده
TransactionCode
هست که معادل شماره یکتا تراکنش در شبکه سراسری شتاب هست

در مورد دریافت شماره کارت و سایر اطلاعات، باید این ویژگی توسعه داده بشه. متاسفانه زمان من برای انجام این گونه کارها محدود هست.
در صورتی که فرصتی پیش بیاد، توسعه داده میشه

موفق باشید
